### DISCUSSION TITLE
GetVersion extremely slow in large repository

### DISCUSSION THREAD

**Discussion by yecarri** [GitTools/GitVersion#3840] [Q&A] [answered]
I have a repository with ~150 branches, of which ~40 are release branches.

Debugging the code I observed there are strategies to calculate the base version not listed in the documentation, in particular 

**GitVersion.VersionCalculation.TrackReleaseBranchesVersionStrategy**

That strategy start jumping back in the commit history and takes ~2 hours to complete. After this time spent, the selected base version is the same as the one selected by TaggedCommitVersionStrategy.

I wonder if TrackReleaseBranchesVersionStrategy can be optimized:

- Select only the greatest release branch, so, for example, in a list of brances named release/1.xxx, release/2.xxx, release/3.0, release/3.0.1 only attempt to get the  base vesion from release/3.0.1
- Limit the number of jumps backward
- Stop looking backward when commit date is older than certain date

All my tests were done in version 5.9.0. Unfortunately I cannot override with a customized gitversion executable due to company's policy.

Any advise from the dev team?

**Comment by arturcic** [comment]
@yecarri I see you're using an old version. So maybe you can try 5.12.0 instead. At the moment we're working on the v6.0 version. It's still in beta but if you are allowed to, you can give it a try.

**Comment by yecarri** [comment]
Hi @arturcic 
Thank you for your response. I checked with 5.12.0 too. Unfortunately the time is exactly the same. I didn't find any change in TrackReleaseBranchesVersionStrategy class after 5.11.0 and the changes in the diff between 5.9.0 and 5.11.0 are mostly aesthetic.

I expected the process to work this way:

1. Select last commit. 
2. If it is tagged with version, terminate.
3. If current commit has 1 parent, select the parent. Goto 2
4. If current commit has 2 or more parents, scan recursively each parent starting in 2.

But what I see is that the algorithm continues looking into ancestors even if it found version tags.

TrackReleaseBranchesVersionStrategy is expensive. It should be called only if ConfigNextVersionVersionStrategy, MergeMessageVersionStrategy, TaggedCommitVersionStrategy, and VersionInBranchNameVersionStrategy failed to get a base version.

And FallbackVersionStrategy should be called only if all the above failed.

I don't see a reason to spend 2 hours jumping over ancestors when I started in a tagged version or I found a tagged version in one or 2 linear jumps. Is there a reason not to apply the strategies this way?

**Comment by arturcic** [comment, marked as answer]
As far as I remember we have some fixes in the 6.x version, so if you don't mind test the latest beta.4 version. There are some breaking changes there, mostly configuration related, but you can give it a try

**Comment by yecarri** [comment]
I'd love to, but I don't have dotnet-8 in my dev machine.

**Reply by arturcic** [reply]
It should work fine with .net6 as well

**Reply by yecarri** [reply]
No, most projects in the solution are multi-target net6, net7, and net8. I could not compile the solution with donet-6 only (or maybe I am just ignorant and there is a way to do it withour editing all the projects)

Good news is that I got authorization to install dotnet-8 and I tried with the head of main, and it ran in 3 minutes and half, which is really good. My team is eager for a solution, I can't ask devops to deploy a beta in our Jenkins prod server, but I'll ask as soon as 6.0.0 is released.

Cheers!

**Reply by arturcic** [reply]
Great you gave it a try locally and have an improved performance.

We might have a couple more betas and then a release. Most probably the first half of the next year

**Comment by HHobeck** [comment]
Great that you have spent so much time. In my opinion 3 minutes is still too long. Actually in 6.x we have touched a lot of code and changed the concepts to address not only how the version numbers are created. For instance the plan is to create an flags enum to give the user the option to decide which strategies are executed or not: 
- https://github.com/GitTools/GitVersion/discussions/3601

Anyway the topic performance test is something which needs to be ensured and this is a task which is open. Would you like to help for instance to create a plan how to test this? Or create example repositories for the performance tests?
